### PR TITLE
Fix metric calculation in examples and setup tests to run on multi-gpu for no_trainer scripts

### LINKED ISSUE #17200: almost all codes that related to generation in examples/pytorch/**_no_trainer.py have bugs
### System Info

```shell
transformers branch main
```

### Wrong Codes in examples/pytorch/**_no_trainer.py

```
for step, batch in enumerate(eval_dataloader):
            with torch.no_grad():
                generated_tokens = accelerator.unwrap_model(model).generate(
                    batch["input_ids"],
                    attention_mask=batch["attention_mask"],
                    **gen_kwargs,
                )

                generated_tokens = accelerator.pad_across_processes(
                    generated_tokens, dim=1, pad_index=tokenizer.pad_token_id
                )
                labels = batch["labels"]
                if not args.pad_to_max_length:
                    # If we did not pad to max length, we need to pad the labels too
                    labels = accelerator.pad_across_processes(batch["labels"], dim=1, pad_index=tokenizer.pad_token_id)

                generated_tokens, labels = accelerator.gather((generated_tokens, labels))
                generated_tokens = generated_tokens.cpu().numpy()
                labels = labels.cpu().numpy()

                if args.ignore_pad_token_for_loss:
                    # Replace -100 in the labels as we can't decode them.
                    labels = np.where(labels != -100, labels, tokenizer.pad_token_id)
                if isinstance(generated_tokens, tuple):
                    generated_tokens = generated_tokens[0]
                decoded_preds = tokenizer.batch_decode(generated_tokens, skip_special_tokens=True)
                decoded_labels = tokenizer.batch_decode(labels, skip_special_tokens=True)

                decoded_preds, decoded_labels = postprocess_text(decoded_preds, decoded_labels)
                # If we are in a multiprocess environment, the last batch has duplicates
                if accelerator.num_processes > 1:
                    if step == len(eval_dataloader):
                        decoded_preds = decoded_preds[: len(eval_dataloader.dataset) - samples_seen]
                        decoded_labels = decoded_labels[: len(eval_dataloader.dataset) - samples_seen]
                    else:
                        samples_seen += decoded_labels.shape[0]
```

here, In the for loop, step will never equal to len(eval_dataloader), so here should be modified to `if step == len(eval_dataloader) - 1`

and

`samples_seen += decoded_labels.shape[0]`

decoded_labels is a list that produced by the postprocess_text(),
list object have no attribute shape.

GLHF

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

just run the examples scripts provided in the readme

### Expected behavior

```shell
samples_seen exceed the dataset size
and also the attribute error
```

### LINKED ISSUE #17214: Bug of the text-classification in examples
### System Info

```shell
When I finetuned the text classification model based on the glue no trainer script, I found a bug in our script.
The URL is below:
https://github.com/huggingface/transformers/blob/main/examples/pytorch/text-classification/run_glue_no_trainer.py#L525

When we use the accelerator for multi-GPU training, the code should transfer from 
if step == len(eval_dataloader)
to 
if step == len(eval_dataloader) -1
Otherwise, it cannot work to filter the last step duplicated samples.
```

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

just run the script with a text classification using multi-GPU accelerator. The problem occurs in the last step for duplicated samples.

### Expected behavior

```shell
I think it should be fixed soon.
```

### PR DESCRIPTION
**Pull request by muellerzr**
# What does this PR do?

This PR fixes all failing tests in a multi-gpu setting for all `no_trainer` example scripts. This includes an issue with when the logger was called before `Accelerator()` was created, adjusting when the conditional to add to the `samples_seen`, and adjusting `samples_seen` to use the length when the labels are just a list instead of a torch tensor.

Because the tests were rewritten to use `accelerate launch` and the new `write_basic_config`, these tests will automatically allocate themselves properly to test on multigpu, gpu, or cpu depending on what environment is available

Fixes # (issue):

Closes https://github.com/huggingface/transformers/issues/17214, https://github.com/huggingface/transformers/issues/17200

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.
@sgugger

### CODE REVIEW

**Review by sgugger — APPROVED** [review]
LGTM as long as the tests pass :-)
Thanks for fixing those examples!
